### DISCUSSION TITLE
Fisheye / hemispherical rendering?

### DISCUSSION THREAD

**Discussion by astromcc** [bezo97/IFSRenderer#29] [Ideas]
I develop animations for a small college planetarium. It would be really helpful if IFSRenderer could render fisheye/hemispherical projections. Is this something you'd be interested in supporting?

**Comment by bezo97** [comment]
Heya, can't promise anything but I'll look into it. It'll take at least 1-2 weeks. Planetarium project sounds great, let me know if you need anything else for the animation.

**Reply by astromcc** [reply]
Wow! That would be fantastic if its something you could develop. No hurry. I only found out about IFSRenderer very recently, so I'm happy to wait. I'd help spread the word to other planetariums which could use your visualization tool for their projects.

**Comment by bezo97** [comment]
This looks easy to implement, won't take much effort. @astromcc Do you have a specific projection in mind? In my initial research I've found that there are [multiple types of fisheye projections](https://wiki.panotools.org/Fisheye_Projection), can you suggest one that would fit your setup the best? I guess your projector's manual probably mentions the specific details. Otherwise I'll go with the simplest one where the projector is at the center of the dome.

**Reply by astromcc** [reply]
A circular fisheye equidistant mapping onto a full frame works best with most planetarium projection systems. The planetarium projection software takes care of the mapping to the dome no matter the location of the physical projectors beneath the hemispherical screen. The corners of the square frame are masked out, physically or in software, so they can be left black. Alternatively, rendering information like frame number, file name, etc. can be written into these corners.

Looking at the wiki page you linked to, this would correspond to the first image on the right rather than the second.

As a point of reference, the full resolution of our projection system is 4k x 4k. Some systems can display 8k x 8k.

**Reply by bezo97** [reply]
Thanks for the insight, this is great. Would you be able to validate an image on your system if I send a test output? Some time in the following days. If all is well then I'll include the feature in the next release.

**Comment by bezo97** [comment]
Here's a test output in 4k, have a look @astromcc. If this is what you had in mind I can finish up the feature and release a new IFSRenderer version sometime next week.
![dome-test-4k](https://github.com/bezo97/IFSRenderer/assets/26039350/05aa906c-4378-4e87-b604-6825c13101d0)

**Comment by astromcc** [comment]
The projection looks correct. To confirm the geometry, are all the rays
parallel? Such as pointing in the z-direction in 3d space?

**Reply by bezo97** [reply]
Yes, they are parallel, looking upwards. Thanks for the confirmation. I'll let you know when the feature is ready next week 👍

**Comment by bezo97** [comment]
Another test - I read that this projection distorts near the edge. Here the blue circles are equally placed from each other but the fisheye distorts the view. 
![image](https://github.com/bezo97/IFSRenderer/assets/26039350/26b3e231-8e95-4622-a010-291ccc6d8b3c)

[Here](https://github.com/bezo97/IFSRenderer/commit/61450b760e7a5869a2d1492dc1832c510bb33463) is the implemented feature which I'll release in a few days.

**Comment by astromcc** [comment]
Hi Zoltan,

Sorry for the delay... Monday was a holiday here at my college, Tuesday was
busy.

I believe I understand the setup of the blue circles. Are they concentric
and aligned to an X-Y plane above the camera / eye position? If so, the
fact that the rings appear closer together near the edge of the projection
is expected. It's not a distortion so much as an effect of perspective.
Looking in the vertical direction, the line of sight is nearly
perpendicular to the X-Y plane so there is little to no foreshortening.
Looking in any direction just above the horizon, the line of sight is
nearly parallel to the X-Y plane. In this second case, the foreshortening
is significant.

**Comment by bezo97** [comment]
This feature is released in [v1.4.0](https://github.com/bezo97/IFSRenderer/releases/tag/v1.4.0) 🥳 

Also @astromcc I've written a small [wiki page](https://github.com/bezo97/IFSRenderer/wiki/Create-a-fulldome-planetarium-production) specifically for this use case. I ask you kindly to please review this with your expertise and extend if needed.  
I can do a follow-up patch in case you find any issues - eager to hear your feedback.

**Reply by vitos1k** [reply]
Thank you for this addition! It's great, we do visualisations for planetariums and this looks great! 
One small note though. Not all domes have their viewers center point in the zenit (center of circle projection). Usually the center point is tilted (30-50 degrees) can you implement additional offset for camera angle for Fisheye and Euirectangular projections?

**Reply by bezo97** [reply]
@vitos1k This may take a while because I have no clue about the math. I've filed an issue for this: #36 Feel free to take it yourself if you're more knowledgeable.
